adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -463,17 +474,103 @@ private void handleConnectionRequestResult(final MqttEndpoint endpoint,
+                final CommandSubscription cmdSub = CommandSubscription.fromTopic(subscription.topicName(),
+                        authenticatedDevice);
+                if (cmdSub != null) {
+                    if (subscription.qualityOfService() != MqttQoS.EXACTLY_ONCE) {
+                        LOG.debug("subscription for [topic:{}, QoS:{}]", subscription.topicName(),
+                                subscription.qualityOfService());
+                        grantedQosLevels.add(subscription.qualityOfService());
+                        endpoint.subscribeAcknowledge(subscribe.messageId(), grantedQosLevels);
+                        createCommandConsumer(endpoint, cmdSub).setHandler(h -> {
+                            if (h.succeeded()) {
+                                h.result().flow(1);
+                                sendTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, -1);
+                            } else {
+                                LOG.error("command consumer could not be created for [tenant:{}, device:{}]: {}",
+                                        cmdSub.getTenant(), cmdSub.getDeviceId(), h.cause());
+                            }
+                        });
+                        // if called from unauthenticated device subscription overwrite the close
+                        // handler with now known tenant and device
+                        if(authenticatedDevice==null) {
+                            endpoint.closeHandler(v -> {
+                                close(endpoint, new Device(cmdSub.getTenant(), cmdSub.getDeviceId()));
+                            });
+                        }
+                    } else {
+                        LOG.error("denied subscription for exactly once");
+                    }
+                }
+            } else {
+                LOG.error("subscription for multiple topics is not allowed");
+            }
+            if (grantedQosLevels.isEmpty()) {
+                grantedQosLevels.add(MqttQoS.FAILURE);
+                endpoint.subscribeAcknowledge(subscribe.messageId(), grantedQosLevels);
+            }
+        }).unsubscribeHandler(unsubscribe -> {
+            if (unsubscribe.topics().size() == 1) {
+                final String topic = unsubscribe.topics().get(0);
+                final CommandSubscription cmdSub = CommandSubscription.fromTopic(topic,
+                        authenticatedDevice);
+                if (cmdSub != null) {
+                    final String tenantId = cmdSub.getTenant();
+                    final String deviceId = cmdSub.getDeviceId();
+                    LOG.debug("unsubscribe for [topic:{}] from [tenant:{}, device:{}]",
+                            topic, tenantId, deviceId);
+                    endpoint.unsubscribeAcknowledge(unsubscribe.messageId());
+                    closeCommandConsumer(tenantId, deviceId);
```

Comment:
I wonder if there might occur a race-condition between a command being sent upstream from an application while we are processing the unsubscribe packet ...

Reply:
FMPOV the command then will not be successful.